RecoBTag/FeatureTools/plugins/ParticleTransformerAK4TagInfoProducer.cc

```diff
@@ -0,0 +1,427 @@
+class ParticleTransformerAK4TagInfoProducer : public edm::stream::EDProducer<> {
+public:
+  explicit ParticleTransformerAK4TagInfoProducer(const edm::ParameterSet&);
+  ~ParticleTransformerAK4TagInfoProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  typedef std::vector<reco::ParticleTransformerAK4TagInfo> ParticleTransformerAK4TagInfoCollection;
+  typedef reco::VertexCompositePtrCandidateCollection SVCollection;
+  typedef reco::VertexCollection VertexCollection;
+
+  void beginStream(edm::StreamID) override {}
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  void endStream() override {}
+
+  const double jet_radius_;
+  const double min_candidate_pt_;
+  const bool flip_;
+
+  edm::EDGetTokenT<edm::View<reco::Jet>> jet_token_;
+  edm::EDGetTokenT<VertexCollection> vtx_token_;
+  edm::EDGetTokenT<SVCollection> sv_token_;
+  edm::EDGetTokenT<edm::ValueMap<float>> puppi_value_map_token_;
+  edm::EDGetTokenT<edm::ValueMap<int>> pvasq_value_map_token_;
+  edm::EDGetTokenT<edm::Association<VertexCollection>> pvas_token_;
+  edm::EDGetTokenT<edm::View<reco::Candidate>> candidateToken_;
+  edm::ESGetToken<TransientTrackBuilder, TransientTrackRecord> track_builder_token_;
+  bool use_puppi_value_map_;
+  bool use_pvasq_value_map_;
+
+  bool fallback_puppi_weight_;
+  bool fallback_vertex_association_;
+
+  bool is_weighted_jet_;
+
+  const double min_jet_pt_;
+  const double max_jet_eta_;
+};
+
+ParticleTransformerAK4TagInfoProducer::ParticleTransformerAK4TagInfoProducer(const edm::ParameterSet& iConfig)
+    : jet_radius_(iConfig.getParameter<double>("jet_radius")),
+      min_candidate_pt_(iConfig.getParameter<double>("min_candidate_pt")),
+      flip_(iConfig.getParameter<bool>("flip")),
+      jet_token_(consumes<edm::View<reco::Jet>>(iConfig.getParameter<edm::InputTag>("jets"))),
+      vtx_token_(consumes<VertexCollection>(iConfig.getParameter<edm::InputTag>("vertices"))),
+      sv_token_(consumes<SVCollection>(iConfig.getParameter<edm::InputTag>("secondary_vertices"))),
+      candidateToken_(consumes<edm::View<reco::Candidate>>(iConfig.getParameter<edm::InputTag>("candidates"))),
+      track_builder_token_(
+          esConsumes<TransientTrackBuilder, TransientTrackRecord>(edm::ESInputTag("", "TransientTrackBuilder"))),
+      use_puppi_value_map_(false),
+      use_pvasq_value_map_(false),
+      fallback_puppi_weight_(iConfig.getParameter<bool>("fallback_puppi_weight")),
+      fallback_vertex_association_(iConfig.getParameter<bool>("fallback_vertex_association")),
+      is_weighted_jet_(iConfig.getParameter<bool>("is_weighted_jet")),
+      min_jet_pt_(iConfig.getParameter<double>("min_jet_pt")),
+      max_jet_eta_(iConfig.getParameter<double>("max_jet_eta")) {
+  produces<ParticleTransformerAK4TagInfoCollection>();
+
+  const auto& puppi_value_map_tag = iConfig.getParameter<edm::InputTag>("puppi_value_map");
+  if (!puppi_value_map_tag.label().empty()) {
```

Comment:
`puppi_value_map` is defined in the fillDescription: how can it be empty?
Unless I miss something, please move all definitions into the class initialization

Reply:
This is related to the following PR : https://github.com/cms-sw/cmssw/pull/40803
Such safety seems to be necessary and we follow the Puppi expert's implementation on the old taggers here

Reply:
Uhm... ok: if you add an empty InputTag in `puppi_value_map` then `use_puppi_value_map_` remains `false`.
Not an ultra-safe procedure, but it is in line with the rest of the code in the BTag area.